Fairmat 2023 (#1271)
nexusformat/definitions: manual/source/classes/contributed_definitions/em-structure.rst

.. _Em-Structure:

=======================
Electron microscopy
=======================

.. index::
   IntroductionEm
   EmAppDef
   EmBC
   EmPartnerClasses

.. _IntroductionEm:

Introduction
############

> Set of data storage objects to describe components of an electron microscope and its eventually available focused-ion beam functionalities. The data storage objects were designed from the perspective of how electron microscopes are used by colleagues in the materials-science-branch of electron microscopy. We realize that the biology-/bio-materials/omics-branch of electron microscopy is eventually in an already more mature state of discussion with respect to data management practices. In what follows the focus is on the usage of electron microscopy in condensed-matter physics, chemical physics of solids, and materials engineering applications. As many of the components of electron microscopes used in the bio-materials communities are the same or at least many components very similar to those used and described in materials science, we are confident that the here presented schema definitions can also inspire discussion and exchange with the bio-materials community in the future.

Reviewer: ```suggestion
Set of data storage objects to describe components of an electron microscope and its eventually available focused-ion beam functionalities. The data storage objects were designed from the perspective of how electron microscopes are used by colleagues in the materials-science branch of electron microscopy. We realize that the biology-/bio-materials/omics-branch of electron microscopy is already in a more mature state of discussion with respect to data management practices. In what follows the focus is on the usage of electron microscopy in condensed-matter physics, chemical physics of solids, and materials engineering applications. As many of the components of electron microscopes used in the biomaterials communities are the same or at least many components are very similar to those used and described in materials science, we are confident that the presented schema definitions can also inspire discussion and exchange with the biomaterials community in the future.
```

Partner consortia in the German National Research Data Infrastructure are here e.g.
NFDI-BioImage, NFDI-Microbiota, NFDI4Health, and e.g. NFDI-Neuro.

> Electron microscopes are functionally very customizable tools: Examples include multi-signal/-modal analyses which are frequently realized as on-the-fly computational analyses, regularly switching between GUI-based instrument control, computational steps, and more and more using high-throughput stream-based processing. Also artificial intelligence methods are increasingly used and become closer interconnected with classical modes of controlling the instrument and perform data processing. A challenge in electron microscopy is that these steps are often executed within commercial integrated control and analysis software. This makes it difficult to keep track of workflows in a technology-partner-agnostic,

Reviewer: ```suggestion
Electron microscopes are functionally very customizable tools: Examples include multi-signal/-modal analyses which are frequently realized as on-the-fly computational analyses, regularly switching between GUI-based instrument control, computational steps, and more and more using high-throughput stream-based processing. Also artificial intelligence methods are increasingly used and are becoming more closely interconnected with classical modes of controlling the instrument and perform data processing. A challenge in electron microscopy is that these steps are often executed within commercial integrated control and analysis software. This makes it difficult to keep track of workflows in a technology-partner-agnostic,
```

Reviewer 2: Edited also in PR1271 now

i.e. interdisciplinary manner.

.. _EmAppDef:

Application Definitions
#######################

> We acknowledge that it can be difficult to agree on a single application definition which is generally enough applicable yet not unnecessary complex and useful for application across a variety of instruments, technology partners, and instrument use cases. In what follows, the proposal conceptualizes first the basic components of an electron microscope and the usual workflow how an electron microscope is used for collecting data with detectors via probing radiation-specimen-matter interaction mechanisms.

Reviewer: ```suggestion
We acknowledge that it can be difficult to agree on a single application definition which is generally enough applicable yet not unnecessarily complex and useful for applications across a variety of instruments, technology partners, and instrument use cases. In what follows, the proposal conceptualizes first the basic components of an electron microscope and the usual workflow of how an electron microscope is used for collecting data with detectors via probing radiation-specimen-matter interaction mechanisms.
```

Reviewer 2: Edited also in PR1271 now

> In summary, scientists place a specimen/sample into the microscope, calibrate the instrument, take measurements, may perform experiments, prepare their specimens with a focused ion beam, calibrate again, and take other measurements, before their session on the instrument ends. In between virtually all of these steps data are collected and stream in from different detectors probing different physical mechanisms of the interaction between electrons or other types of radiation with the specimen.

Reviewer 2: Edited also in PR1271 now

A microscope session ends with the scientist removing the specimen from the instrument or parking it so that the next user can start a session. Occasionally, service technicians perform calibrations and maintenance which also can be described as session on the microscope. Next, we wrote base classes to describe these steps and events.

    :ref:`NXem`:
        A general application definition which explores the possibilities of electron microscopes.

.. _EmBC:

Base Classes
############

The following base classes are proposed to support modularizing the storage of pieces of information:

    :ref:`NXaberration_model`, :ref:`NXaberration_model_ceos`, :ref:`NXaberration_model_nion`, :ref:`NXaberration`:
        Base classes to describe procedures and values for the calibration of aberrations based on either CEOS or Nion.

    :ref:`NXaperture_em`:

>         A base class to describe an aperture.

Reviewer 2: Agree, NXslit and or NXaperture would match:
However, there is an inconsistence currently:
https://manual.nexusformat.org/classes/base_classes/NXaperture.html#nxaperture
vs
https://manual.nexusformat.org/classes/base_classes/NXslit.html#nxslit
an EM aperture maybe qualifies for be more than simple so NXaperture as a compromise to use but that is deprecated ...

>         A base class to describe an aperture.

Reviewer 2: Unless that inconsistence is resolved let's keep NXaperture_em, although if it is really an NXaperture only for em, then we should revert the symbol for the base class i.e NXem_aperture to although I have to say we have a constant struggle with the ordering of the field-specific "prefix" like "em" and the symbol for the concept specifier here "aperture", I agree though that the em specific content could equally go into NXem and then one would just say (NXaperture) or (NXslit) (in yaml notation) but it is really the fact that aperture is deprecated because of which I am hesitant specifically because either way fact is we have aperture (several of them in EM) and we want need to describe them and these apertures have similarities with other components called aperture useful not only for EM

>         A base class to describe an aperture.

Reviewer: Also noted as in an issue in #1231.  I think  "This group is deprecated, use NXslit instead" overrides what appears to be left over text "A simple slit. For more complex geometries, NXaperture should be used." in NXslit.  Therefore I think my question remains.  What is missing from NXslit that requires a new NXaperture_em base class?  Or maybe better stated, could we pull enough stuff into NXslit from NXaperture to make NXaperture_em unneeded?

>         A base class to describe an aperture.

Reviewer: Argument from @sanbrock that convinced me: making this base class avoids a lot of copy-pasting between many application definitions.  Resolving.  Thanks!

    :ref:`NXpump`:
        A base class to describe details about a pump in an instrument.

    :ref:`NXscanbox_em`:
        A base class to represent the component of an electron microscope which realizes a controlled deflection (and eventually shift, blanking, and/or descanning) of the electron beam to illuminate the specimen in a controlled manner. This can be used to document the scan pattern.

    :ref:`NXspectrum_set`:
        Base class and specializations comparable to NXimage_set but for storing spectra. Specialized base classes should use controlled vocabulary items as prefixes such as **eels** electron energy loss spectroscopy, **xray** X-ray spectroscopy (EDS/STEM, EDX, SEM/EDX, SEM/EDS), **auger** Auger spectroscopy, or **cathodolum** for cathodoluminescence spectra.

    :ref:`NXstage_lab`:
        As it was mentioned for atom probe microscopy, this is a base class to describe the stage/specimen holder which offers place for the documentation of the small-scale laboratory functionalities which modern stages of electron microscopes frequently offer.

.. _EmPartnerClasses:

Partner application definitions
###############################

> A partner application definition is considered an application definition which stores data and metadata which are relevant for a given experiment but have usually only few connections to the detailed description of the workflow and experiment which motivates to granularize these pieces of information in an own application definition. In fact, one limitation of application definitions in NeXus is that they define a set of constraints on their graph of controlled concepts and terms. If we take for example diffraction experiments with an electron microscope it is usually the case that (diffraction) patterns are collected in the session at the microscope but all scientifically relevant conclusions are drawn later, i.e. through post-processing these data. These numerical and algorithmic steps define computational workflows were data from the application definitions such as NXem are used as input but many additional concepts and constraints may apply without any need for changing constraints on fields or groups of NXem. If we were to modify NXem for these cases, NXem would likely combinatorially diverge as every different combination of required constraints demands having an own but almost similar application definition. For this reason, the concept of partner application definition which have fields/links where specifically relevant sources of information are connected to e.g. NXem.

Reviewer: ```suggestion
A partner application definition is considered an application definition that stores data and metadata that are relevant for a given experiment but have usually only few connections to the detailed description of the workflow and experiment,  motivating the granularization of these pieces of information in their own application definition. In fact, one limitation of application definitions in NeXus is that they define a set of constraints on their graph of controlled concepts and terms. If we take for example diffraction experiments with an electron microscope it is usually the case that (diffraction) patterns are collected in the session at the microscope but all scientifically relevant conclusions are drawn later, i.e. through post-processing these data. These numerical and algorithmic steps define computational workflows where data from the application definitions such as NXem are used as input but many additional concepts and constraints may apply without any need for changing constraints on fields or groups of NXem. If we were to modify NXem for these cases, NXem would likely combinatorially diverge as every different combination of required constraints would demand having their own but almost similar application definition. For this reason, the concept of partner application definitions is introduced here that have fields/links where specifically relevant sources of information are connected to them, e.g. NXem.
```

Reviewer 2: Edited also in PR1271 now

> More consolidation through the use of NXsubentry classes should be considered in the future.

Reviewer: This pull request is adding a new "term of art" to NeXus, that of "Partner application definitions".  I'm not totally following from the description though.  @sanbrock can you restate what is going on with these?  Are there enough to justify a new NeXus term of art?

Reviewer 2: It is good that you raise the question here whether it would need a new technical term. I am not convinced that this is really needed, my point was just that there are practical cases where to describe a workflow of measurement and analyses steps there are substantiated reasons for not having all pieces of information granularized into one appdef / base class but use a number of appdef i.e. there is some logical connection between different appdefs at least in how these may be used in certain fields.
I follow your suggestion and drop the term, instead I thought hard about how to connect different sets of granularized pieces of information in the field of EM.
The idea is essentially as follows and this is reflected in the refactored EM proposal:
NXem is the head appdef which specifies the arity and existence constraints which groups fields etc from a base class should be used, NXem_base is a proposal how one can use a base class to build a tree to relevant building blocks and base class instances to become together as a set useful and at least meaty enough to keep data, metadata and components of an EM and EM research. Initially we proposed base classes like image_secondary electron but there is nothing which makes an image secondary electron different from image backscattered electron, only the fact that what one sees in the image is different. Realizing this we now go for a concept/base class image and rather hook this base class into a deep base class e.g. NXem_eds which builds a tree to detail e.g. typical pieces of information worth to be documented for EDS energy-dispersive X-ray spectroscopy.
This NXem_eds base class is hooked-in in turn then into NXem_base.
The main idea behind NXem_base is thus to offer a template that still can be used when the only difference between a usage of an appdef is that several fields are required or optional because otherwise one would strictly speaking need as many appdefs as one has differly constraints trees how base classes are composed in EM research.

Reviewer 2: Edited also in PR1271 now

    :ref:`NXpump`:
        A base class to describe details about a pump in an instrument.

    :ref:`NXscanbox_em`:
        A base class to represent the component of an electron microscope which realizes a controlled deflection (and eventually shift, blanking, and/or descanning) of the electron beam to illuminate the specimen in a controlled manner. This can be used to document the scan pattern.

    :ref:`NXspectrum_set`:
        Base class and specializations comparable to NXimage_set but for storing spectra. Specialized base classes should use controlled vocabulary items as prefixes such as **eels** electron energy loss spectroscopy, **xray** X-ray spectroscopy (EDS/STEM, EDX, SEM/EDX, SEM/EDS), **auger** Auger spectroscopy, or **cathodolum** for cathodoluminescence spectra.

    :ref:`NXstage_lab`:
        As it was mentioned for atom probe microscopy, this is a base class to describe the stage/specimen holder which offers place for the documentation of the small-scale laboratory functionalities which modern stages of electron microscopes frequently offer.

> .. _EmPartnerClasses:

Reviewer 2: Edited
